pkg/controllers/managementuser/resourcequota/resource_quota_sync.go

```diff
@@ -119,68 +117,65 @@ func (c *SyncController) CreateResourceQuota(ns *corev1.Namespace) (runtime.Obje
 		return ns, err
 	}
 
-	projectLimit, _, err := getProjectResourceQuotaLimit(ns, c.ProjectLister)
-	if err != nil {
-		return ns, err
-	}
-
-	var quotaSpec *corev1.ResourceQuotaSpec
-	if projectLimit != nil {
-		quotaSpec, err = c.getNamespaceResourceQuota(ns)
-		if err != nil {
-			return ns, err
-		}
-	}
-
-	quotaToUpdate, err := c.getResourceQuotaToUpdate(ns)
+	requestedQuotaLimit, newQuotaSpec, err := c.deriveRequestedResourceQuota(ns)
 	if err != nil {
 		return ns, err
 	}
 
 	operation := "none"
 	if existing == nil {
-		if quotaSpec != nil {
+		if newQuotaSpec != nil {
 			operation = "create"
 		}
 	} else {
-		if quotaSpec == nil {
+		if newQuotaSpec == nil {
 			operation = "delete"
-		} else if quotaToUpdate != "" || !apiequality.Semantic.DeepEqual(existing.Spec.Hard, quotaSpec.Hard) {
+		} else if requestedQuotaLimit != nil || !apiequality.Semantic.DeepEqual(existing.Spec.Hard, newQuotaSpec.Hard) {
 			operation = "update"
 		}
 	}
 
 	var updated *corev1.Namespace
-	var isFit bool
+	var operationErr error
 	switch operation {
 	case "create":
-		isFit, updated, err = c.validateAndSetNamespaceQuota(ns, quotaToUpdate)
+		isFit, updated, exceeded, err := c.validateAndSetNamespaceQuota(ns, &v32.NamespaceResourceQuota{Limit: *requestedQuotaLimit})
 		if err != nil {
 			return updated, err
 		}
 		if !isFit {
-			// create default "all 0" resource quota
-			quotaSpec, err = getDefaultQuotaSpec()
+			// Create a quota with zeros only for overused resources.
+			limit, err := zeroOutResourceQuotaLimit(requestedQuotaLimit, exceeded)
+			if err != nil {
+				return updated, err
+			}
+
+			newQuotaSpec, err = convertResourceLimitResourceQuotaSpec(limit)
 			if err != nil {
 				return updated, err
 			}
 		}
-		err = c.createDefaultResourceQuota(ns, quotaSpec)
+		operationErr = c.createResourceQuota(ns, newQuotaSpec)
 	case "update":
-		isFit, updated, err = c.validateAndSetNamespaceQuota(ns, quotaToUpdate)
-		if err != nil || !isFit {
+		isFit, updated, exceeded, err := c.validateAndSetNamespaceQuota(ns, &v32.NamespaceResourceQuota{Limit: *requestedQuotaLimit})
+		if err != nil {
 			return updated, err
 		}
-		err = c.updateResourceQuota(existing, quotaSpec)
+		if !isFit {
+			operationErr = fmt.Errorf("update quota doesn't fit due to the following resource requests: [%v]", format.ResourceList(exceeded))
+			break
+		}
+		operationErr = c.updateResourceQuota(existing, newQuotaSpec)
 	case "delete":
-		err = c.deleteResourceQuota(existing)
+		operationErr = c.deleteResourceQuota(existing)
 	}
 
 	if updated == nil {
 		updated = ns
 	}
 
-	if err != nil {
+	if operationErr != nil {
+		logrus.Tracef("Failed to perform operation %q on namespace %q: %v", operation, ns.Name, operationErr)
 		return updated, err
```

Comment:
This is now returning the wrong error
```suggestion
		return updated, operationErr
```